explorer/interpreter/action_stack.cpp

```diff
@@ -194,6 +209,10 @@ auto ActionStack::Spawn(std::unique_ptr<Action> child, RuntimeScope scope)
 auto ActionStack::ReplaceWith(std::unique_ptr<Action> replacement)
     -> ErrorOr<Success> {
   std::unique_ptr<Action> old = todo_.Pop();
+  CARBON_CHECK(FinishActionKindFor(old->kind()) ==
+               FinishActionKindFor(replacement->kind()))
+      << "Can't replace action " << *old << " with action " << *replacement
+      << " that has a different kind of result";
```

Comment:
```suggestion
      << " that has a different FinishActionKind";
```

"kind of result" could be incorrectly inferred to mean a different Carbon type, I think it'd be better be precise.

Though TBH I think you could remove this line and not lose much, since the CARBON_CHECK condition seems reasonably clear.